runner/utils/driver/docker/docker.go

```diff
@@ -0,0 +1,252 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/sirupsen/logrus"
+	"github.com/txn2/txeh"
+
+	"github.com/spaceuptech/space-cloud/runner/model"
+)
+
+type docker struct {
+	client *client.Client
+}
+
+func NewDockerDriver() (*docker, error) {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		logrus.Errorf("error creating docker module instance in docker in docker unable to initialize docker client - %v", err)
+		return nil, err
+	}
+	return &docker{client: cli}, nil
+}
+
+// ApplyService creates containers for specified service
+func (d *docker) ApplyService(service *model.Service, token string) error {
+	ctx := context.Background()
+	// remove containers if already exits
+	if err := d.DeleteService(service.ID); err != nil {
+		logrus.Errorf("error applying service in docker unable delete existing containers - %v", err)
+		return err
+	}
+
+	// todo check host file overiding concern
+	// client for CRUD operation on host file
+	// default location /etc/hosts
+	hostFile, err := txeh.NewHostsDefault()
+	if err != nil {
+		logrus.Errorf("error applying service in docker unable to load host file with suitable default - %v", err)
+		return err
+	}
+
+	for _, task := range service.Tasks {
+		// todo get image
+		// for now images have been created locally but not uploaded on docker hub
+		out, err := d.client.ImagePull(ctx, task.Docker.Image, types.ImagePullOptions{})
+		if err != nil {
+			return err
+		}
+		io.Copy(os.Stdout, out)
+
+		// TODO check docker runtime field in host config
+		// expose ports of docker container as specified in task
+		exposedPorts := map[nat.Port]struct{}{}
+		for _, port := range task.Ports {
+			portString := strconv.Itoa(int(port.Port))
+			portWithProtocol := nat.Port(fmt.Sprintf("%s/%s", portString, port.Protocol))
+			exposedPorts[portWithProtocol] = struct{}{}
+		}
+
+		// set environment variables of docker container
+		task.Env["URL"] = fmt.Sprintf("http://service.artifact.svc.cluster.local:4122")
+		task.Env["TOKEN"] = fmt.Sprintf("%s", token) // todo token
```

Comment:
You need to set these only if the `Runtime` field of the service config is `code`. The token to pass here should be generated on the spot. You'll need to put the following claims in this:
- id: service id
- project: project id
- version: version id